### DISCUSSION TITLE
Queries to another database as resources

### DISCUSSION THREAD

**Discussion by pdekraker-epa** [ckan/ckan#6003] [Q&A]
I am brainstorming some plans to connect another database with CKAN, essentially with some resources being queries of that database, and am curious if there is any guidance or examples of something similar. 

My base use case could be pretty simple, with queries to the database being fairly static, essentially requesting data from a test we ran with only a couple different variations. I believe an IBlueprint and IResourceController could be set up to make the proper connections and serve out the data in a manner comparable to datastore.dump. 

This leads me to question if there is a way to leverage the datastore to manage this. I read IDatastoreBackend as being an interface to enable replacing Postgres with another database, but not enabling connection to multiple different databases based on resource attributes. Is this interpretation correct?

Alternatively could I implement methods / blueprints overriding datastore_search and datastore.dump directing them to the other DB when applicable and falling back for "normal" resources? Part of the reason I would consider this over a separate implementation would be to be able to leverage datastore based resource views. 

If you have a better idea about how to handle this I would love to hear it...

**Comment by jqnatividad** [comment]
I have a similar use case, and wanted to just connect to an existing high-volume, transaction database, without having to physically replicate the data in the CKAN datastore.
Played around with [foreign data wrappers](https://wiki.postgresql.org/wiki/Foreign_data_wrappers) with some success.

This is one of my motivations for creating the request to add `datastore_register` to the Datastore API (https://github.com/ckan/ckan/issues/5855), as I wanted to create a Foreign Table connection to a read-only materialized view in the transaction database, as this will allow me to create a near real-time connection to the data whilst skipping the ETL tango of exporting, pushing/xloading the data into the Datastore, and still be able to leverage the different resource view plugins and CKAN's resource-level mechanisms for access control, activity stream, extra metadata, etc.

**Reply by pdekraker-epa** [reply]
Thanks for this, I am far from a Database expert so getting a handle on the capabilities and limitations is great. My primary goal would be using resource views which in turn use the datastore api. Other resources can create a compatible interface, and should be able to use them too. The other DB to which I would be connecting happens to be Oracle based, and like you I would only want read-only access. Using a wrapper like you mentioned seems like a more encompassing solution than trying to squeeze in a special case. 

In thinking about via FDW or otherwise, maybe the the datastore could become something of a middleware, connecting multiple backends and distributing or merging results.  Given that there are other DB technologies, having a capability to store some resources in Postgres and others in MongoDB or another NoSQL could be a good path in the future.

**Reply by jqnatividad** [reply]
That's what attracted me to the FDW approach - it's a generic, lightweight solution that leverages PostgreSQL's capabilities, in a way that _should_ be relatively transparent to the existing DataStore extension.

In my case, I was connecting to SQL Server.

I did some initial experiments with https://github.com/cphsolutionslab/ckanext-dataproxy, but the extension is not maintained. Ultimately, I ended up shifting priorities and focused on [revamping datatablesview instead, which was recently finished](https://github.com/ckan/ckan/pull/5952).

Perhaps, now that we're using GitHub Discussions, we can use this thread to socialize requirements gathering?

**Reply by pdekraker-epa** [reply]
@jqnatividad I am circling back to this topic, particularly foreign data wrappers. I have spent some time learning and have an application in mind using a FDW to pull in data from an external api for which I would like to have the datatore table views / visualizations available. I am thinking of making an IDatastoreBackend extending the default backend, but enabling creation of foreign tables. At a high level this seems straightforward, but I'm wondering if you could help identify any potential issues / pitfalls. Looking at the default postgres backend I see code for triggers, and added columns ( _id & _full_text ), but am unsure what requirements they would add for writing such an extension.

**Reply by jqnatividad** [reply]
@pdekraker-epa I think the most pragmatic, practical approach is to introduce `datastore_register` ( https://github.com/ckan/ckan/issues/5855).

In that way, you don't need to create a new datastore backend and it just works transparently with the existing postgres backend as the routing to the foreign table is handled on the database layer by PostgreSQL.

`datastore_register` will also fully leverage Postgres' features (like setting up partitioned table, optimizing indices, etc) that your DBA can do, instead of depending on CKAN to do all the DDL setup.

I was actually hoping to start on it last year when I was doing datastore-related work, but the funding for it dried up before I got around to it.

**Comment by anuveyatsu** [comment]
@pdekraker-epa maybe not exactly what you're looking for but we have an extension for using BigQuery for DataStore here: https://github.com/datopian/ckanext-datastore-bigquery
